File: backend/check.py
"""Run with python3 backend/check.py. Uses an isolated, disposable database."""
from concurrent.futures import ThreadPoolExecutor
from io import BytesIO
import json
import sqlite3
from pathlib import Path
from tempfile import TemporaryDirectory

from server import create_app, validate_event, RequestError


def request(app, method="GET", path="/mqsf/api/events", data=None, origin=None, content_type="application/json"):
    payload = json.dumps(data).encode() if data is not None else b""
    environ = {"REQUEST_METHOD": method, "PATH_INFO": path, "CONTENT_TYPE": content_type,
               "CONTENT_LENGTH": str(len(payload)), "wsgi.input": BytesIO(payload),
               "wsgi.url_scheme": "http", "HTTP_HOST": "localhost:8030"}
    if origin:
        environ["HTTP_ORIGIN"] = origin
    captured = []
    body = b"".join(app(environ, lambda status, headers: captured.append((int(status[:3]), dict(headers)))))
    assert len(captured) == 1, "Each request must return exactly one response"
    status, headers = captured[0]
    return status, json.loads(body) if body and "application/json" in headers["Content-Type"] else body, headers


with TemporaryDirectory() as directory:
    database = Path(directory) / "events.sqlite3"
    app = create_app(database, allowed_origins=("https://munich-quantum-software.github.io",))
    event = {"date": "2026-10-14", "start": "10:00", "end": "11:00", "title": 'Meetup <script>alert("x")</script>',
             "description": "Bring a laptop.\nEveryone is welcome.", "audience": "Developers & researchers", "organizers": "Alex & Sam"}
    assert request(app)[1] == {"events": [], "demo": False}
    assert request(app, "POST", data=event, origin="https://unrelated.example")[0] == 403
    assert request(app, "POST", data=event, content_type="text/plain")[0] == 415
    status, data, headers = request(app, "POST", data=event, origin="https://munich-quantum-software.github.io")
    assert status == 201 and headers["Access-Control-Allow-Origin"] == "https://munich-quantum-software.github.io"
    first = data["event"]
    assert first["title"] == event["title"] and first["version"] == 1
    assert set(first) == set(event) | {"id", "version", "updated_at"}, "Store only event data and revision metadata"
    assert first["organizers"] == event["organizers"]
    second_client = create_app(database)
    assert request(second_client)[1]["events"] == [first], "Edits must persist across clients and restarts"
    assert request(app, "POST", data={**event, "start": "10:30", "end": "11:30"})[0] == 201, "Overlaps are allowed"
    path = "/mqsf/api/events/" + first["id"]
    with ThreadPoolExecutor(max_workers=2) as pool:
        results = list(pool.map(lambda title: request(app, "PUT", path, {**first, "title": title})[0], ["Editor one", "Editor two"]))
    assert sorted(results) == [200, 409], "Concurrent edits must not overwrite each other"
    latest = next(e for e in request(app)[1]["events"] if e["id"] == first["id"])
    assert latest["version"] == 2
    assert request(app, "DELETE", path, {"version": 1})[0] == 409
    for changes in [{"date": "2026-10-16"}, {"start": "11:00"}, {"end": "09:00"}, {"start": "24:00"},
                    {"title": " "}, {"title": "x" * 121}, {"description": None}, {"audience": 5},
                    {"organizers": " "}, {"organizers": "x" * 201}, {"start": "07:59"}]:
        assert request(app, "POST", data={**event, **changes})[0] == 400, changes
    assert request(app, "PUT", path, {**latest, "version": True})[0] == 400
    assert request(app, "POST", data={**event, "description": "x" * 25000})[0] == 413
    bounds = {"days": [{"date": "2026-10-14", "start": "09:00", "end": "17:00"}]}
    assert validate_event(event, bounds)["start"] == "10:00"
    try:
        validate_event({**event, "start": "08:00"}, bounds)
        raise AssertionError("Published hours must be enforced on the server")
    except RequestError as error:
        assert error.status == 400
    assert request(app, "POST", data={**event, "start": "18:00", "end": "21:00"})[0] == 201, "Networking has no published cutoff"
    status, data, _ = request(app, "PUT", path, {**latest, "organizers": "Taylor"})
    assert status == 200 and data["event"]["organizers"] == "Taylor"
    latest = data["event"]
    assert request(app, "DELETE", path, {"version": latest["version"]})[0] == 200
    assert request(app, "PUT", path, latest)[0] == 404
    assert request(app, "GET", "/mqsf/../backend/server.py")[0] == 404
    assert request(app, "GET", "/mqsf/.data/events.sqlite3")[0] == 404
    assert request(app, "GET", "/mqsf")[0] == 308
    for mount in ["/mqsf", "/side-events"]:
        assert request(app, "GET", mount + "/")[0] == 200
        assert request(app, "GET", mount + "/app.mjs")[0] == 200
        assert request(app, "GET", mount + "/api/events")[0] == 200
    assert b'href="./side-events/"' in request(app, "GET", "/")[1]
    for resource in ["/styles.css", "/script.js", "/assets/images/brand/favicon.png", "/assets/images/brand/mqsf-logo.svg"]:
        assert request(app, "GET", resource)[0] == 200
    assert request(app, "GET", "/assets/../backend/server.py")[0] == 404
    assert request(app, "OPTIONS", origin="https://munich-quantum-software.github.io")[0] == 204
    legacy_db = Path(directory) / "legacy.sqlite3"
    with sqlite3.connect(legacy_db) as connection:
        connection.execute("CREATE TABLE events (id TEXT PRIMARY KEY, date TEXT, start TEXT, end TEXT, title TEXT, description TEXT, audience TEXT, version INTEGER, updated_at TEXT)")
        connection.execute("INSERT INTO events VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)", tuple(first[key] for key in ("id", "date", "start", "end", "title", "description", "audience", "version", "updated_at")))
    migrated = request(create_app(legacy_db))[1]["events"][0]
    assert migrated == {**first, "organizers": ""}, "Adding organizers must preserve existing events"
    assert request(create_app(legacy_db))[1]["events"][0] == migrated, "Migration is safe to repeat"
    seeded_db = Path(directory) / "seeded.sqlite3"
    seeded = create_app(seeded_db, seed_examples=True)
    examples = request(seeded)[1]
    assert examples["demo"] is False and len(examples["events"]) == 5
    assert all(e["title"].endswith(" (example)") and e["organizers"] == "Example organizer" for e in examples["events"])
    assert request(create_app(seeded_db, seed_examples=True))[1] == examples, "Seeding must not duplicate or replace existing events"
    before = request(app)[1]
    assert request(create_app(database, seed_examples=True))[1] == before, "Seeding must leave an existing calendar untouched"
print("Calendar API checks passed: persistence, organizers, migration, concurrent edits, deletion, hours, navigation, and path isolation.")

File: cloudflare/check.mjs
// An isolated local D1 database: no account or production data is used.
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import { getPlatformProxy, unstable_splitSqlQuery } from "wrangler";
import { Miniflare, convertV4MiniflareOptions } from "miniflare";
import worker from "./worker.mjs";
import { notifyChanges } from "./notifications.mjs";
import { restoreSQL } from "./moderate.mjs";

const platform = await getPlatformProxy({ configPath: "cloudflare/wrangler.jsonc", persist: false, remoteBindings: false });
const { env } = platform;
const origin = "https://munich-quantum-software.github.io";
async function request(method = "GET", path = "/api/events", data, from = origin) {
  const response = await worker.fetch(new Request(`https://calendar.example${path}`, {
    method, headers: { Origin: from, ...(data !== undefined ? { "Content-Type": "application/json" } : {}) },
    ...(data !== undefined ? { body: JSON.stringify(data) } : {}),
  }), env);
  return { status: response.status, headers: response.headers, data: response.status === 204 ? null : await response.json() };
}
async function apply(file, db = env.DB) {
  const sql = await readFile(new URL(file, import.meta.url), "utf8");
  await db.batch(unstable_splitSqlQuery(sql).map(sql => db.prepare(sql)));
}
try {
  await apply("./migrations/0001_events.sql");
  await apply("./migrations/0003_private_contacts_and_history.sql");
  assert.deepEqual((await request()).data, { events: [], demo: false });
  const draft = { date: "2026-10-14", start: "10:00", end: "11:15", title: "Meetup <b>plain text</b>",
    description: "Bring a laptop.\nAll welcome.", audience: "Developers", organizers: "Alex & Sam", contact_email: "private@example.test" };
  assert.equal((await request("POST", "/api/events", draft, "https://unrelated.example")).status, 403);
  const preflight = await request("OPTIONS");
  assert.equal(preflight.status, 204);
  assert.equal(preflight.headers.get("Access-Control-Allow-Origin"), origin);
  const created = await request("POST", "/api/events", draft);
  assert.equal(created.status, 201);
  const first = created.data.event, path = `/api/events/${first.id}`;
  const { contact_email, ...publicDraft } = draft;
  assert.deepEqual(first, { ...publicDraft, id: first.id, version: 1, updated_at: first.updated_at });
  assert.deepEqual((await request()).data.events, [first]);
  assert.equal((await request("POST", "/api/events", { ...draft, start: "10:30", end: "11:30" })).status, 201);
  const simultaneous = await Promise.all(["One", "Two"].map(title => request("PUT", path, { ...first, title })));
  assert.deepEqual(simultaneous.map(r => r.status).sort(), [200, 409]);
  const latest = simultaneous.find(r => r.status === 200).data.event;
  assert.equal(latest.version, 2);
  assert.equal((await request("DELETE", path, { version: 1 })).status, 409);
  for (const change of [{ contact_email: "" }, { contact_email: null }, { contact_email: "not-an-email" }, { contact_email: "a@example.test\r\nBcc:x" }, { organizers: " " }, { organizers: "x".repeat(201) }, { audience: null }, { date: "2026-10-16" },
    { start: "24:00" }, { start: "07:59" }, { start: "11:15" }, { end: "09:00" }]) {
    assert.equal((await request("POST", "/api/events", { ...draft, ...change })).status, 400);
  }
  assert.equal((await request("PUT", path, { ...latest, version: true })).status, 400);
  assert.equal((await request("POST", "/api/events", { ...draft, description: "x".repeat(20001) })).status, 413);
  assert.equal((await request("POST", "/api/events", { ...draft, start: "18:00", end: "21:00" })).status, 201);
  const edited = await request("PUT", path, { ...latest, organizers: "Taylor" });
  assert.equal(edited.data.event.organizers, "Taylor");
  assert.equal((await env.DB.prepare("SELECT contact_email FROM events WHERE id=?").bind(first.id).first()).contact_email, contact_email, "Blank edits preserve the hidden address");
  const replaced = await request("PUT", path, { ...edited.data.event, contact_email: "replacement@example.test" });
  assert.equal(replaced.status, 200);
  assert.equal(JSON.stringify(replaced.data).includes("contact_email"), false);
  assert.equal((await request("DELETE", path, { version: replaced.data.event.version })).status, 200);
  const history = (await env.DB.prepare("SELECT * FROM event_changes WHERE event_id=? ORDER BY rowid").bind(first.id).all()).results;
  assert.deepEqual(history.map(row => row.action), ["created", "updated", "updated", "updated", "deleted"], "Conflicts must not create history");
  assert.equal(JSON.parse(history.at(-2).before_json).contact_email, contact_email);
  assert.equal(JSON.parse(history.at(-1).before_json).contact_email, "replacement@example.test");
  assert.equal(history.at(-1).after_json, null);
  for (const privatePath of ["/api/history", "/api/event_changes", "/api/events/" + first.id]) {
    assert.ok([404, 405].includes((await request("GET", privatePath)).status));
  }
  assert.equal(JSON.stringify((await request()).data).includes("contact_email"), false);
  assert.equal((await request("PUT", path, edited.data.event)).status, 404);
  assert.equal((await request("GET", "/worker.mjs")).status, 404);
  await apply("./migrations/0002_example_events.sql");
  const seeded = (await request()).data.events;
  const examples = seeded.filter(event => event.title.endsWith(" (example)"));
  assert.equal(examples.length, 5);
  const example = examples[0];
  await request("PUT", `/api/events/${example.id}`, { ...example, title: "Participant's updated title" });
  const beforeReseed = (await request()).data.events;
  await apply("./migrations/0002_example_events.sql");
  assert.deepEqual((await request()).data.events, beforeReseed, "Seed migration must preserve participants' edits");
  const deletion = history.at(-1);
  const restored = await env.DB.prepare(restoreSQL(deletion.id, "before", 0)).first();
  assert.equal(restored.id, first.id);
  assert.equal(restored.version, replaced.data.event.version + 1, "Restoration must invalidate all old versions");
  assert.equal((await env.DB.prepare("SELECT contact_email FROM events WHERE id=?").bind(first.id).first()).contact_email, "replacement@example.test");
  assert.equal(await env.DB.prepare(restoreSQL(deletion.id, "before", 0)).first(), null, "Do not overwrite an event recreated since review");
  assert.equal(await env.DB.prepare(restoreSQL(deletion.id, "before", restored.version - 1)).first(), null);
  const reverted = await env.DB.prepare(restoreSQL(history[0].id, "after", restored.version)).first();
  assert.equal(reverted.version, restored.version + 1);
  assert.equal((await env.DB.prepare("SELECT title FROM events WHERE id=?").bind(first.id).first()).title, draft.title);
  assert.equal((await request("PUT", path, replaced.data.event)).status, 409, "Pre-deletion clients cannot overwrite a restored event");

  // A failed journal write must roll back the public mutation as well.
  const beforeFailure = (await request()).data.events;
  await env.DB.prepare("CREATE TRIGGER reject_history BEFORE INSERT ON event_changes BEGIN SELECT RAISE(ABORT, 'test failure'); END").run();
  assert.equal((await request("POST", "/api/events", draft)).status, 503);
  assert.deepEqual((await request()).data.events, beforeFailure);
  await env.DB.prepare("DROP TRIGGER reject_history").run();

  const realFetch = globalThis.fetch, deliveries = [];
  const notifyEnv = { ...env, DISCORD_WEBHOOK_URL: "https://discord.com/api/webhooks/123/test-only" };
  try {
    globalThis.fetch = async (url, options) => {
      assert.equal(new URL(url).search, "?wait=true");
      assert.equal(options.redirect, "manual");
      const payload = JSON.parse(options.body.get("payload_json"));
      assert.deepEqual(payload.allowed_mentions, { parse: [] });
      assert.ok(payload.content.length <= 2000);
      const snapshot = JSON.parse(await options.body.get("files[0]").text());
      deliveries.push(snapshot);
      return Response.json({ id: "discord-message" });
    };
    await notifyChanges(env); // No configured secret: leave the outbox untouched.
    assert.equal(deliveries.length, 0);
    await Promise.all([notifyChanges(notifyEnv), notifyChanges(notifyEnv)]);
    assert.equal(deliveries.length, 6);
    assert.equal(new Set(deliveries.map(d => d.change_id)).size, 6, "Concurrent drains must claim different changes");
    assert.equal(deliveries[0].after.contact_email, contact_email);
    let attempts = 0;
    globalThis.fetch = async () => { attempts++; return Response.json({ retry_after: 120, message: "Do not store response bodies" }, { status: 429 }); };
    await notifyChanges(notifyEnv);
    assert.equal(attempts, 1);
    const pending = await env.DB.prepare("SELECT * FROM event_changes WHERE notify_error IS NOT NULL").first();
    assert.equal(pending.notified_at, null);
    assert.equal(pending.notify_error, "Discord HTTP 429");
    assert.ok(pending.notify_after >= Math.floor(Date.now() / 1000) + 119);
    await env.DB.prepare("UPDATE event_changes SET notified_at='test' WHERE id!=?").bind(pending.id).run();
    await notifyChanges(notifyEnv);
    assert.equal(attempts, 1, "Respect retry backoff");
    await env.DB.prepare("UPDATE event_changes SET notify_after=0 WHERE id=?").bind(pending.id).run();
    globalThis.fetch = async () => { attempts++; return Response.json({ id: "retried-message" }); };
    await notifyChanges(notifyEnv);
    assert.equal(attempts, 2);
    assert.ok((await env.DB.prepare("SELECT notified_at FROM event_changes WHERE id=?").bind(pending.id).first()).notified_at);
    await assert.rejects(notifyChanges({ ...env, DISCORD_WEBHOOK_URL: "https://unrelated.example/webhook" }));
  } finally { globalThis.fetch = realFetch; }
  // Run the actual outbound request in workerd: its Fetch API differs from Node's.
  let runtimeDeliveries = 0, redirectTest = true;
  const runtime = new Miniflare(convertV4MiniflareOptions({
    modules: true, compatibilityDate: "2026-09-25", d1Databases: ["DB"],
    bindings: { DISCORD_WEBHOOK_URL: notifyEnv.DISCORD_WEBHOOK_URL },
    script: await readFile(new URL("./notifications.mjs", import.meta.url), "utf8")
      + "\nexport default { async fetch(request, env) { await notifyChanges(env); return new Response('done'); } };",
    outboundService: async request => {
      runtimeDeliveries++;
      assert.equal(new URL(request.url).hostname, "discord.com");
      const form = await request.formData();
      assert.deepEqual(JSON.parse(form.get("payload_json")).allowed_mentions, { parse: [] });
      assert.equal(JSON.parse(await form.get("files[0]").text()).after.contact_email, contact_email);
      return redirectTest ? new Response(null, { status: 302, headers: { Location: "https://unrelated.example/" } }) : Response.json({ id: "runtime-message" });
    },
  }));
  try {
    const db = await runtime.getD1Database("DB");
    await apply("./migrations/0001_events.sql", db);
    await apply("./migrations/0003_private_contacts_and_history.sql", db);
    await db.prepare("INSERT INTO events(id,date,start,end,title,description,audience,organizers,contact_email,updated_at) VALUES(?,?,?,?,?,?,?,?,?,?)")
      .bind(crypto.randomUUID(), draft.date, draft.start, draft.end, draft.title, draft.description, draft.audience, draft.organizers, contact_email, new Date().toISOString()).run();
    await runtime.dispatchFetch("http://localhost/");
    assert.equal(runtimeDeliveries, 1, "Never follow webhook redirects with private data");
    assert.equal((await db.prepare("SELECT notify_error FROM event_changes").first()).notify_error, "Discord HTTP 302");
    redirectTest = false;
    await db.prepare("UPDATE event_changes SET notify_after=0").run();
    await runtime.dispatchFetch("http://localhost/");
    assert.equal(runtimeDeliveries, 2);
    assert.ok((await db.prepare("SELECT notified_at FROM event_changes").first()).notified_at);
  } finally { await runtime.dispose(); }
  console.log("Cloudflare D1 checks passed: private contacts, atomic history, Discord outbox, retries, CRUD, validation, concurrent edits, CORS, hours, and example seeds.");
} finally { await platform.dispose(); }
